phoenix-core/src/test/java/org/apache/phoenix/index/PrepareIndexMutationsForRebuildTest.java

```diff
@@ -0,0 +1,732 @@
+ */
+
+package org.apache.phoenix.index;
+
+import org.apache.hadoop.hbase.Cell;
+import org.apache.hadoop.hbase.CellUtil;
+import org.apache.hadoop.hbase.KeyValue;
+import org.apache.hadoop.hbase.client.Delete;
+import org.apache.hadoop.hbase.client.Mutation;
+import org.apache.hadoop.hbase.client.Put;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.phoenix.coprocessor.IndexRebuildRegionScanner;
+import org.apache.phoenix.hbase.index.IndexRegionObserver;
+import org.apache.phoenix.jdbc.PhoenixConnection;
+import org.apache.phoenix.query.BaseConnectionlessQueryTest;
+import org.apache.phoenix.query.QueryConstants;
+import org.apache.phoenix.schema.PTable;
+import org.apache.phoenix.schema.PTableKey;
+import org.apache.phoenix.util.SchemaUtil;
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.assertEquals;
+
+public class PrepareIndexMutationsForRebuildTest extends BaseConnectionlessQueryTest {
+    private static String ROW_KEY = "k1";
+    private static String TABLE_NAME = "dataTable";
+    private static String INDEX_NAME = "idx";
+
+    class SetupInfo {
+        public IndexMaintainer indexMaintainer;
+        public PTable pDataTable;
+    }
+
+    /**
+     * Get the index maintainer and phoenix table definition of data table.
+     * @param tableName
+     * @param indexName
+     * @param columns
+     * @param indexColumns
+     * @param pk
+     * @param includeColumns
+     * @return
+     * @throws Exception
+     */
+    private SetupInfo setup(String tableName,
+                            String indexName,
+                            String columns,
+                            String indexColumns,
+                            String pk,
+                            String includeColumns) throws Exception {
+        try(Connection conn = DriverManager.getConnection(getUrl())) {
+
+            String fullTableName = SchemaUtil.getTableName(SchemaUtil.normalizeIdentifier(""), SchemaUtil.normalizeIdentifier(tableName));
+            String fullIndexName = SchemaUtil.getTableName(SchemaUtil.normalizeIdentifier(""), SchemaUtil.normalizeIdentifier(indexName));
+
+            // construct the data table and index based from the parameters
+            String str1 = String.format("CREATE TABLE %1$s (%2$s CONSTRAINT pk PRIMARY KEY (%3$s)) COLUMN_ENCODED_BYTES=0",
+                    fullTableName,
+                    columns,
+                    pk);
+            conn.createStatement().execute(str1);
+
+            String str2 = String.format("CREATE INDEX %1$s ON %2$s (%3$s)",
+                    fullIndexName,
+                    fullTableName,
+                    indexColumns);
+            if (!includeColumns.isEmpty())
+                str2 += " INCLUDE (" + includeColumns + ")";
```

Comment:
why not use string buffer instead?

Reply:
Are you taking about the performance difference between them. String buffer is more efficient when doing such concatenating, i will prefer to use it for production code. Since it is unit test code, I would prefer to keep it in a more readable way.

Reply:
Not sure what makes using string buffer not so readable as it is used in several tests in the project. :)